frontend/src/services/promotions.ts

```diff
@@ -1,98 +1,119 @@
-}
+  const { cuisineType, dayOfWeek, discountType, serviceOptions } = filters;
 
-function sortDistance(promotions: Promotion[]) {
-  // TODO: sort based on restaurant distance
+  let promotions = [...arr];
+  if (filters.cuisineType.length > 0) {
+    promotions = filterCuisineType(promotions, cuisineType);
+  }
+  if (filters.dayOfWeek.length > 0) {
+    promotions = filterDayOfWeek(promotions, dayOfWeek);
+  }
+  if (filters.discountType.length > 0) {
+    promotions = filterDiscountType(promotions, discountType);
+  }
+  if (filters.serviceOptions.length > 0) {
+    promotions = filterServiceOptions(promotions, serviceOptions);
+  }
   return promotions;
 }
 
-function sortPopularity(promotions: Promotion[]) {
-  // TODO: sort based on number of saves
-  return promotions;
+function filterCuisineType(promotions: Promotion[], filters: string[]) {
+  let result: Promotion[] = [];
+  for (const key of filters) {
+    const filtered = promotions.filter(({ cuisine }) => {
+      const sanitized = cuisine.replace("\\s", "_");
+      return sanitized.toUpperCase() === key.toUpperCase();
+    });
+    result = [...result, ...filtered];
+  }
+  return result;
 }
 
-function sortRating(promotions: Promotion[]) {
-  // TODO: sort based on number of ratings
-  return promotions;
+function filterDayOfWeek(promotions: Promotion[], filters: string[]) {
+  let result: Promotion[] = [];
+  for (const key of filters) {
+    const filtered = promotions.filter(({ schedules }) => {
+      return schedules.find(({ dayOfWeek }) => dayOfWeek === key);
+    });
+    result = [...result, ...filtered];
+  }
+  return result;
 }
 
-export function filterPromotions(
-  promotions: Promotion[],
-  filter: FilterBy
-): Promotion[] {
-  if (enumContainsValue(Category, filter)) {
-    return filterCategory(promotions, filter as Category);
-  } else if (enumContainsValue(CuisineType, filter)) {
-    return filterCuisineType(promotions, filter as CuisineType);
-  } else if (enumContainsValue(DaysOfWeek, filter)) {
-    return filterDaysOfWeek(promotions, filter as DaysOfWeek);
-  } else if (enumContainsValue(DiscountType, filter)) {
-    return filterDiscountType(promotions, filter as DiscountType);
-  } else if (enumContainsValue(ServiceOptions, filter)) {
-    return filterServiceOptions(promotions, filter as ServiceOptions);
+function filterDiscountType(promotions: Promotion[], filters: string[]) {
+  let result: Promotion[] = [];
+  for (const key of filters) {
+    const filtered = promotions.filter((promotion) => {
+      const {
+        discount: { discountType },
+      } = promotion;
+      return discountType === key.substring(0, 1);
+    });
+    result = [...result, ...filtered];
   }
-  return promotions;
+  return result;
 }
 
-function filterCategory(promotions: Promotion[], filter: Category) {
-  return promotions.filter(
-    ({ category }) => category.replace("\\s", "_").toUpperCase() === filter
-  );
+// TODO: filter based on restaurant's service options
```

Comment:
Should we create issues for all these `TODO` comments and link them here in the code? That way BE and FE can keep better track of what we still need to do

Reply:
Ya good idea! I'll set these up as issues

Reply:
Added the sorting functionality to this issue: https://github.com/ubclaunchpad/foodies/issues/99

Reply:
Thanks!